spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/session/SessionAutoConfiguration.java

```diff
@@ -78,6 +85,29 @@
 			SessionRepositoryFilterConfiguration.class })
 	static class ServletSessionConfiguration {
 
+		private final ServerProperties serverProperties;
+
+		ServletSessionConfiguration(ServerProperties serverProperties) {
+			this.serverProperties = serverProperties;
+		}
+
+		@Bean
+		@ConditionalOnMissingBean({ CookieSerializer.class,
+				HeaderHttpSessionIdResolver.class })
```

Comment:
After taking a second look I think this would be better off being conditional on missing `CookieSerializer` and conditional on `CookieHttpSessionIdResolver`, because that's the component using the cookie serializer we're configuring here. Otherwise this would switch off when `HeaderHttpSessionIdResolver` bean is present, but not when users provide their custom `HttpSessionIdResolver` implementation.

Thoughts?

Reply:
Those are the gotchas that I was referring to in #15139 and that we started to discuss in #10440 before we went with the `SessionCookieConfig`-based approach. I'll need to spend some time refreshing my memory.

Reply:
After giving this some more thought, I've added a commit with custom condition that kicks in if either of the following is true:

* no `HttpSessionIdResolver` and `CookieSerializer` beans are registered - this is the default state with Spring Session
* `CookieHttpSessionIdResolver` bean is registered but `CookieSerializer` is not - this is when users configure a custom `CookieHttpSessionIdResolver` but they don't register the `CookieSerializer`

This covers the case (a quite valid one) when users provide their custom `HttpSessionIdResolver` implementation. That shouldn't result in `CookieSerializer` being configured. There's a test in second commit that covers this scenario.

IMO the options are either to have this composite condition approach, or abandon the case when users provide their own `CookieHttpSessionIdResolver` but not `CookieSerializer` - one might argue that since `CookieHttpSessionIdResolver` itself is `final` and `CookieSerializer` is the only thing can be configure on it that it's an edge case.

Do you have any thoughts on this @wilkinsona?

Reply:
Thanks, @vpavic. The scenario that you've described is the one that I needed to refresh my memory about. I like the custom condition approach that you've implemented. This looks ready to merge to me. Thanks for the PR.